### ISSUE TITLE
No progress callback for Forge download

### ISSUE THREAD

**Issue by LangArthur** [FlowArg/FlowUpdater#15]
Hi Flow,
I'm using your library to create a Minecraft launcher with Forge. I implemented the update method for FlowUpdater progress callback to display download step in my UI. It works like a charm while downloading vanilla minecraft.
But as soon as it starts to download Forge / mods, there is not possibility to keep track of the download progress.

Is it a missing feature ? Or is it me who missed something ?

**Comment by FlowArg** [comment]
Hi,
It really depends on the way FlowUpdater installs Forge/Fabric actually. Indeed we currently download the Forge installer (or Fabric as well) and run it in a new Java process and wait for the installation to be completed. So we have no control over the installation of the mod loader so we can't figure out what the installer is currently downloading.

Notwithstanding there are two solutions:
- Keep the installer system as it is currently and get some infos from the installer output. We'll be able to gather some information but not enough to do what we want.
- Completly change how the installation of the mod loaders works by bundling directly in FlowUpdater the installation of the mod loaders. In this case, we have a complete control over the modloaders' installation. However, there are still some cons: If the mod loader changes how the installer works, we'll have to adapt our code too. The Forge installer code is also a mess to edit, I've already passed hours to adapt it to how FlowUpdater can launch it. No problem with Fabric. So maybe we should keep the installation of Fabric based mod loaders as it is now and we should change only the forge installation ?

TL;DR:
- We can make some tricks or hacks to gather some info easily but not sure it's very worth it.
- We can also change how the installation of the mod loader works. Pros: full control over the installation ; no longer zip extraction to set up the forge installer. Cons: It'll take many time to implement. Maybe it will cause some incompatibilities in the future. If we do that, I guess I'll do it only for Forge since Fabric based mod loaders installers are fast enough.

Your question is really interesting because it asks a back spot of FlowUpdater. I've plan for months (and maybe years now haha) to change how FlowUpdater installs mod loaders and especially Forge. But I didn't have the motivation and the time to do that.

**Comment by LangArthur** [comment]
Hi,

Sorry for the late answer, I was a bit busy these days :)

As you say, it might be a lot of work for not that much result.
My use case was just about displaying feedback for user, which can be less detailed.

However, about the different solutions you proposed, changing the full mod loader installation seems to me the "best" solution, on a technical view, but need times and maintenance (I am not an expert, but I understood that forge API change regularly), which will surely not fit your situation. 

I close the issue with this comment, considering there is no easy fix, but it remains something to improve for the future, for somebody motivated :)
